firewood/src/merkle/stream.rs

```diff
@@ -396,6 +396,140 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+        // at the same time as immutable access to `merkle`.
+        let Self { state, merkle } = &mut *self;
+
+        match state {
+            PathIteratorState::Exhausted => None,
+            PathIteratorState::Iterating {
+                matched_key,
+                unmatched_key,
+                address,
+            } => {
+                let node = match merkle.get_node(*address) {
+                    Ok(node) => node,
+                    Err(e) => return Some(Err(e)),
+                };
+
+                let partial_path = match node.inner() {
+                    NodeType::Branch(branch) => &branch.path,
+                    NodeType::Leaf(leaf) => &leaf.path,
+                    _ => unreachable!("extension nodes shouldn't exist"),
+                };
+
+                let node_key: Box<[u8]> = matched_key
+                    .iter()
+                    .chain(partial_path.iter())
+                    .copied()
+                    .collect();
+
+                let (comparison, unmatched_key) =
+                    compare_partial_path(partial_path.iter(), unmatched_key);
+
+                match comparison {
+                    Ordering::Less | Ordering::Greater => {
+                        self.state = PathIteratorState::Exhausted;
+                        Some(Ok((node_key, node)))
+                    }
+                    Ordering::Equal => match node.inner() {
+                        NodeType::Branch(branch) => {
+                            let Some(next_unmatched_key_nibble) = unmatched_key.next() else {
+                                // There's no more key to match. We're done.
+                                self.state = PathIteratorState::Exhausted;
+                                return Some(Ok((node_key, node)));
+                            };
+
+                            #[allow(clippy::indexing_slicing)]
+                            let Some(child) = branch.children[next_unmatched_key_nibble as usize] else {
+                                // There's no child at the index of the next nibble in the key.
+                                // The node we're seeking isn't in the trie.
+                                self.state = PathIteratorState::Exhausted;
+                                return Some(Ok((node_key, node)));
+                            };
+
+                            *matched_key = node_key
+                                .iter()
+                                .chain(once(&next_unmatched_key_nibble))
+                                .copied()
+                                .collect();
+                            *address = child;
+
+                            Some(Ok((node_key, node)))
+                        }
+                        NodeType::Leaf(_) => {
+                            self.state = PathIteratorState::Exhausted;
+                            Some(Ok((node_key, node)))
+                        }
```

Comment:
Nit, I would put this at the top. I tend to find code easier to read if the degenerate case is written first (when it's possible).